doc/nbfc.1.in
.nh
.TH "NBFC" "1" "APRIL 2026" "Notebook FanControl @VERSION@"
.SH NAME
.PP
nbfc \- Notebook FanControl client

.SH DESCRIPTION
.PP
This program is used to control the NoteBook FanControl service.

.SH SYNOPSIS
.PP
.SS SERVICE CONTROL COMMANDS
.B nbfc
.RB { start " | " stop " | " restart " | " status " | " config " | " set " | " sensors }
.RI [ OPTIONS ]
.SS MISCELLANEOUS COMMANDS
.B nbfc
.RB { update " | " rate\-config " | " test\-config " | " acpi\-dump " | " get\-model\-name " | " xml2json " | " support " | " warranty " | " faq " | " help }
.RI [ OPTIONS ]

.SH OPTIONS
.PP
.BR \-h ", " \-\-help
.RS
Show help message and exit.
.RE

.PP
.B \-\-version
.RS
Show program's version number and exit.
.RE

.SH COMMANDS
.PP

.B start
.RI [ OPTIONS ]
.RS
Start the service.

.BR \-r ", " \-\-read\-only
.RS
Start in read\-only mode.
.RE
.RE

.B stop
.RS
Stop the service.
.RE

.B restart
.RI [ OPTIONS ]
.RS
Restart the service.

.BR \-r ", " \-\-read\-only
.RS
Start in read\-only mode.
.RE
.RE

.B status
.RI [ OPTIONS ]
.RS
Show the service status.

.BR \-a ", " \-\-all
.RS
Show service and fan status (default).
.RE

.BR \-s ", " \-\-service
.RS
Show service status.
.RE

.BR \-f ", " \-\-fan
.I FAN INDEX
.RS
Show fan status.
.RE

.BR \-w ", " \-\-watch
.I SECONDS
.RS
Show status periodically.
.RE

.BR \-j ", " \-\-json
.RS
Use JSON output.

If specified, all other options are ignored.
.RE
.RE

.B config
.RI [ OPTIONS ]
.RS
List or apply configs.

.BR \-l ", " \-\-list
.RS
List all available configs.
.RE

.BR \-r ", " \-\-recommend
.RS
List configs with a similar notebook model name.
.RE

.BR \-s ", " \-\-set
.I CONFIG
.RS
Set a config.
.RE

If
.I CONFIG
is
.BR auto ,
the service will attempt to automatically select a matching configuration.

.I CONFIG
is the configuration filename without the
.B .json
extension, not the value of the
.B NotebookModel
field inside the configuration file.
.RE

.B set
.RI [ OPTIONS ]
.RS
Control fan speed.

.BR \-a ", " \-\-auto
.RS
Set fan speed to profile-based control.
.RE

.BR \-s ", " \-\-speed
.I PERCENT
.RS
Set fan speed to PERCENT.
.RE

.BR \-f ", " \-\-fan
.I FAN INDEX
.RS
Fan index (zero based). If not given, all available fans are set to speed.
.RE
.RE

.B sensors
.RB { list " | " set " | " show }
.RI [ OPTIONS ]
.RS
Configure fan sensors.

.B list
.RS
List all available sensors and their temperature files.
.RE

.B show
.RS
Show all available fans and their sensor configuration.
.RE

.B set
.RI [ OPTIONS ]
.RS
Configure which sensors are attached to a fan.

.BR \-f ", " \-\-fan
.I FAN INDEX
.RS
Specifies the fan to configure.
.RE

.BR \-s ", " \-\-sensor
.I SENSOR
.RS
Sensor to add. Can be specified multiple times.
.RE

.BR \-a ", " \-\-algorithm
.I ALGORITHM
.RS
Algorithm
.RB ( Average ", " Min ", " Max )
.RE

.B \-\-force
.RS
Force applying a sensor if it could not be found.
.RE
.RE
.RE

.B update
.RI [ OPTIONS ]
.RS
Download new configuration files from the internet.

Downloaded files will be stored in
.IR /var/lib/nbfc/configs .

.BR \-p ", " \-\-parallel
.I NUM
.RS
Set the number of parallel downloads (default: 10).
.RE

.BR \-q ", " \-\-quiet
.RS
Set quiet mode.
.RE
.RE

.B rate\-config
.RI [ FILE "] [" OPTIONS ]
.RS
Rates a configuration by analyzing whether it appears safe to execute on the
current system.

By default, the DSDT and all SSDTs are read from /sys/firmware/acpi/tables.

.BR \-H ", " \-\-full\-help
.RS
Show help on how to interpret the results.
.RE

.BR \-a ", " \-\-all
.RS
Rate all available configuration files.
.RE

.BR \-b ", " \-\-bad
.RS
List only bad configs that are otherwise omitted.
.RE

.BR \-d ", " \-\-dsdt
.I FILE
.RS
Use an alternative DSDT file. Can be specified multiple times.
.RE

.BR \-D ", " \-\-dsdt\-dir
.I DIRECTORY
.RS
Use an alternative DSDT directory.
.RE

.BR \-f ", " \-\-fan\-count
.I NUMBER
.RS
Limit output to configurations with the specified number of fans.
.RE

.BR \-i ", " \-\-input
.I FILE
.RS
Read configuration files from
.IR FILE .
.RE

.BR \-j ", " \-\-json
.RS
Use JSON output.
.RE

.BR \-m ", " \-\-min-score
.I SCORE
.RS
Set minimum rating threshold for displaying configurations (default: 9.0).
.RE

.BR \-n ", " \-\-no\-download
.RS
Don't download configuration rules from the repository.
.RE

.BR \-r ", " \-\-rules
.I FILE
.RS
Use an alternative rules file.
.RE

.BR \-u ", " \-\-unverified
.RS
Normally, only registers belonging to an
.B EmbeddedControl
operation region are taken into account when rating a configuration.
With this option, registers from
.B SystemMemory
operation regions that may be accessible through the embedded controller are also considered.
Their accessibility through the embedded controller cannot be guaranteed.
.RE

.BR \-q ", " \-\-quiet
.RS
Do not print register ratings.
.RE

.B \-\-print\-rules
.RS
Print configuration rating rules.
.RE
.RE

.B test\-config
.RB { run " | " evaluate }
.RI [ OPTIONS ]
.RS
Test configurations automatically.

This command should only be used in conjunction with
.BR "nbfc rate-config" .

For example:
.RS
.B sudo nbfc rate-config -aqq | sudo nbfc test-config run
.RE

It should
.B not
be used to test arbitrary configurations.

.B run
.RI [ OPTIONS ]
.RS
Reads a list of configuration files from an input file (default:
.BR stdin ).

Configurations will be loaded in read\-only mode.

The notebook's fans will be spinned up by stressing the CPU/GPU while reading
back the fan speed for each configuration.

The result will be written to
.BR nbfc.test-config.result.json .

.BR \-i ", " \-\-input
.I FILE
.RS
Specify input file.
.RE

.BR \-o ", " \-\-output
.I FILE
.RS
Specify output file.
.RE

.BR \-c ", " \-\-cpu
.I NUMBER
.RS
Specify number of CPU workers.
.RE

.BR \-g ", " \-\-gpu
.I NUMBER
.RS
Specify number of GPU workers (default:
.BR "no workers" ).
.RE

.BR \-I ", " \-\-interval
.I SECONDS
.RS
Set sample interval (default:
.BR 0.5 ).
.RE

.BR \-t ", " \-\-threshold
.I DELTA
.RS
Stop when temperature difference falls below
.I DELTA
(default:
.BR 2.0 ).
.RE

.BR \-b ", " \-\-break
.I SECONDS
.RS
Stop after the temperature difference has remained below
.I DELTA
for
.I SECONDS
consecutive seconds (default:
.BR 10.0 ).
.RE
.RE

.B evaluate
.RI [ OPTIONS ]
.RS
Evaluates a previously generated result file. Default:
.BR nbfc.test-config.result.json .

.BR \-i ", " \-\-input
.I FILE
.RS
Specify input file.
.RE
.RE
.RE

.B acpi\-dump
.RB { registers " | " ec\-registers " | " methods " | " dsl " | " map }
.RI [ OPTIONS ]
.RS
Dumps information of your ACPI tables.

By default, the DSDT and all SSDTs are read from /sys/firmware/acpi/tables.

.B registers
.RS
List all available registers.
.RE

.B ec\-registers
.RB [ \-u | \-\-unverified ]
.RS
List all available embedded controller registers.

Normally, only registers belonging to an
.B EmbeddedControl
operation region are printed.
If
.BR \-u | \-\-unverified
is specified, also registers from
.B SystemMemory
operation regions that may be accessible through the embedded controller are printed.
Their accessibility through the embedded controller cannot be guaranteed.
.RE

.B methods
.RS
List all available ACPI methods.
.RE

.B dsl
.RS
Disassemble your ACPI tables.
.RE

.B map
.RB [ \-u | \-\-unverified ]
.RS
Generate a map file that can be used by
.BR "ec_probe -m|--map" .
The map contains detected registers with their names and addresses as well as
ACPI method names.

Normally, only registers belonging to an
.B EmbeddedControl
operation region are printed.
If
.BR \-u | \-\-unverified
is specified, also registers from
.B SystemMemory
operation regions that may be accessible through the embedded controller are printed.
Their accessibility through the embedded controller cannot be guaranteed.
.RE

.BR \-d ", " \-\-dsdt
.I FILE
.RS
Use an alternative DSDT file. Can be specified multiple times.
.RE

.BR \-D ", " \-\-dsdt\-dir
.I DIRECTORY
.RS
Use an alternative DSDT directory.
.RE

.BR \-j ", " \-\-json
.RS
Use JSON output.
.RE
.RE

.B support
.RI [ OPTIONS ]
.RS
Show how to support the project.

.B \-\-upload\-firmware
.RS
Upload your notebook firmware without prompting.
.RE

.B \-\-print\-command
.RS
Print command for manual firmware upload.
.RE

.B \-\-create\-archive
.I FILE
.RS
Creates a compressed
.B tar.gz
archive containing information required for support and hardware analysis.

The archive includes:

.IP \(bu 2
The output of
.B nbfc get-model-name
.IP \(bu 2
The output of
.B nbfc sensors list
.IP \(bu 2
Firmware DSDT
.RI ( /sys/firmware/acpi/tables/DSDT )
.IP \(bu 2
Firmware SSDTs
.RI ( /sys/firmware/acpi/tables/SSDT* )
.RE
.RE

.B get\-model\-name
.RS
Print out the notebook's model name.
.RE

.B xml2json
.I FILE
.RS
Convert an XML configuration file to JSON.

This command can be used to convert configurations from the original NBFC project
into the NBFC\-Linux format.
.RE

.B warranty
.RS
Show warranty.
.RE

.B faq
.RS
Show NBFC FAQ manual page.
.RE

.B help
.RS
Show help.
.RE

.SH FILES
.PP

.I @SYSCONFDIR@/nbfc/nbfc.json
.RS
The main configuration file for nbfc_service. See
.BR nbfc_service.json (5)
for further details.
.RE

.I @RUNSTATEDIR@/nbfc_service.pid
.RS
File containing the PID of current running nbfc_service.
.RE

.I @RUNSTATEDIR@/nbfc_service.socket
.RS
Socket file of nbfc_service.
.RE

.I @DATADIR@/nbfc/configs/*.json
.RS
Configuration files for various notebook models. See
.BR nbfc_service.json (5)
for further details.
.RE

.I @DATADIR@/nbfc/model_support.json
.RS
Model support database. This file maps notebook model names to configuration files.
.RE

.I /var/lib/nbfc/configs/*.json
.RS
Configuration files downloaded from the internet. Files in this directory take precedence over those in
.IR @DATADIR@/nbfc/configs .
.RE

.I /var/lib/nbfc/model_support.json
.RS
Model support database downloaded from the internet. Both
.I /var/lib/nbfc/model_support.json
and
.I @DATADIR@/nbfc/model_support.json
are taken into account.
.RE

.I /var/lib/nbfc/state.json
.RS
State file of nbfc_service. This holds the current fan speeds.
.RE

.SH EXIT STATUS
.RS
.IP 0
Everything fine
.IP 1
Generic error
.IP 2
Command-line error
.IP 3
Initialization error
.IP 5
Fatal error (memory allocation failure)
.RE

.SH BUGS
.PP
Report bugs to https://github.com/nbfc\-linux/nbfc\-linux

.SH AUTHOR
.PP
Benjamin Abendroth (braph93@gmx.de)

.SH SEE ALSO
.PP
.BR nbfc_service (1),
.BR nbfc_service.json (5),
.BR ec_probe (1),
.BR fancontrol (1)


CHANGELOG.md
# Changelog

## nbfc-linux-0.5.4 (2026-10-10)

- Increased numerical limits for `/sys/class/hwmon[0-9]+` files
  - Hardware with more sensors work now
  - Hardware with higher hwmon numbers work now

- Added `nbfc test-config` to refine the output of `nbfc rate-config -a`.
  The new command calculates the correlation between CPU/GPU temperature
  and the reported fan speed for each configuration.

- Added `nbfc reset-ec` (internal command only) for resetting the embedded
  controller after the service dies unexpectedly. This command is used
  in the systemd service file.

- Added support for CPU/GPU temperatures in `ec_probe monitor` and
  `ec_probe graph`.

- Added `ec_probe evaluate` for calculating the correlation between CPU/GPU
  temperature and EC registers. This command operates on output files
  generated by `ec_probe monitor --report=FILE`.

- Added the `-m|--map FILE` option to all `ec_probe` commands dealing with
  registers or ACPI methods. Using a map file, registers can be written by
  their name and ACPI methods can be abbreviated. The map file can be generated
  by `nbfc acpi-dump map`.
  Using `-m|--map` also allows the shell to auto-complete registers / methods.

- Added `nbfc status --json` option for querying the service in a
  machine-readable format.

- Removed `nbfc config -a|--apply` option.
  A notice will be printed saying that `nbfc config -s|--set` should be used
  instead.

- Added `nbfc acpi-dump ec-registers -u|--unverified` option.
  This option includes registers that are not officially exposed through
  the embedded controller. They use the `SystemMemory` operation region
  instead of `EmbeddedControl`.

- Added options to `nbfc rate-config`:
  - `-i|--input=FILE`: Read configuration file list from `FILE`
  - `-q|--quiet`: Suppress rating information
  - `-f|--fan-count=NUM`: Filter configurations by fan count
  - `-b|--bad`: Show bad configurations
  - `-u|--unverified`: See `nbfc acpi-dump ec-registers -u|--unverified`

- Changes in model configuration:
  - `CriticalTemperature` can now be set per `FanConfiguration`.
  - Added `ReadWords`/`WriteWords` to allow read and write sizes to be
    configured independently instead of using `ReadWriteWords`.

- `nbfc_service` now checks if the hardware it is running on has been changed.

- `nbfc_service` now limits the number of read failures reported per fan (less log spam).

## nbfc-linux-0.5.3 (2026-06-02)

- Added firmware fingerprinting support to improve automatic configuration
  matching and compatibility verification.

## nbfc-linux-0.5.2 (2026-05-10)

- Added `nbfc xml2json` command
- Support returning non-integer results in `ec_probe acpi_call`

## nbfc-linux-0.5.1 (2026-04-26)

- Added Lua helper functions for handling packages and buffers returned
  from ACPI methods:

  - `acpi_call_raw(method)`: Executes an ACPI method and returns the result as a string
  - `acpi_get_int(result, path)`: Extracts an integer from `result` located at `path`

## nbfc-linux-0.5.0 (2026-04-20)

- Added support for executing Lua code.

  In `FanConfiguration`:
    - `ReadLuaCode`
    - `WriteLuaCode`
    - `ResetLuaCode`

  In `RegisterWriteConfiguration`:
    - `LuaCode` (requires `WriteMode` set to `Lua`)
    - `ResetLuaCode` (requires `ResetWriteMode` set to `Lua`)

  The following functions are exposed to Lua:
    - `acpi_call(method)`: Executes an ACPI method.
    - `ec_read(register)`: Reads a byte from the embedded controller.
    - `ec_read_word(register)`: Reads two bytes and combines them into a single value.
    - `ec_write(register, value)`: Writes a byte to the embedded controller.
    - `ec_write_word(register, value)`: Writes a 16-bit value to a two-byte register.

## nbfc-linux-0.4.0 (2026-03-14)

- Reduced memory usage of the service from ~280KB to ~230KB.

- Added `nbfc acpi-dump` command

  This command allows to dump information from your ACPI firmware.

  Subcommands:
    - `nbfc acpi-dump methods`: Dumps the list of ACPI methods
    - `nbfc acpi-dump registers`: Dumps the list of all ACPI registers
    - `nbfc acpi-dump ec-registers`: Dumps the list of EC registers
    - `nbfc acpi-dump dsl`: Dumps the disassembled ACPI firmware

- Added `nbfc rate-config` command

  This command rates configuration files by analyzing whether they appear
  safe to use on the current system.

- Added `nbfc support` command

  This command shows how to support the project. It also allows to upload
  the current notebook firmware to the NBFC-Linux server by running
  `nbfc support --upload-firmware`.

## nbfc-linux-0.3.20 (2025-06-06)
- Added support for calling ACPI methods

  NBFC-Linux can now call ACPI methods to read or write the fan speed.
  ACPI methods are also supported in RegisterWriteConfigurations.
  This is useful if the fan cannot be controlled via simple EC register
  manipulations.

  In `FanConfiguration`:
    - `ReadAcpiMethod`: Read the fan speed using an ACPI method
    - `WriteAcpiMethod`: Write the fan speed using an ACPI method
    - `ResetAcpiMethod`: Reset the fan using an ACPI method

  In `RegisterWriteConfiguration`:
    - `AcpiMethod` (requires `WriteMode` set to `Call`)
    - `ResetAcpiMethod` (requires `ResetWriteMode` set to `Call`)

  The `ec_probe` tool now supports calling methods using the `acpi_call` command.

  This requires the `acpi_call` module to be installed.

## nbfc-linux-0.3.18 (2025-05-24)
- New model configuration files

## nbfc-linux-0.3.17 (2025-05-22)
- Using state file to store fan speeds

  Previously, `TargetFanSpeeds` were stored in the main configuration file,
  which required rewriting the file whenever fan speeds changed.

  This could lead to conflicts if the configuration was being edited
  simultaneously via the GUI.

  Fan speeds are now stored in a separate state file to avoid these issues
  and ensure safe, concurrent updates.

## nbfc-linux-0.3.16 (2025-05-10)
- Added `nbfc sensors` command

## nbfc-linux-0.3.15 (2025-05-08)

completion/nbfc.py
#!/usr/bin/python3

import argparse

from argparse_shell_complete import argparse_mod

NBFC_CONFIGS_DIR = '@DATADIR@/nbfc/configs'

argp = argparse.ArgumentParser(prog='nbfc', description='NoteBook FanControl CLI Client')

argp.add_argument('--version', action='version', version='nbfc 0.1.6')

subp = argp.add_subparsers(description='commands')

# =============================================================================
# Command 'start'
# =============================================================================
cmdp = subp.add_parser('start', help='Start the service')

cmdp.add_argument('-r', '--read-only', action='store_true',
                  help='Start in read-only mode')

# =============================================================================
# Command 'stop'
# =============================================================================
cmdp = subp.add_parser('stop', help='Stop the service')

# =============================================================================
# Command 'restart'
# =============================================================================
cmdp = subp.add_parser('restart', help='Restart the service')

cmdp.add_argument('-r', '--read-only', action='store_true',
                  help='Restart in read-only mode')

# =============================================================================
# Command 'status'
# =============================================================================
cmdp = subp.add_parser('status', help='Show the service status')
agrp = cmdp.add_mutually_exclusive_group(required=True)

agrp.add_argument('-a', '--all', action='store_true',
                  help='Show service and fan status (default)')

agrp.add_argument('-s', '--service', action='store_true',
                  help='Show service status')

agrp.add_argument('-f', '--fan', type=int, metavar='FAN INDEX',
                  help='Show status of fan (zero based)')

cmdp.add_argument('-w', '--watch',
                  help='Show status periodically', type=float, metavar='SECONDS')

# =============================================================================
# Command 'config'
# =============================================================================
cmdp = subp.add_parser('config', help='List or apply configs')
agrp = cmdp.add_mutually_exclusive_group(required=True)

agrp.add_argument('-l', '--list', action='store_true',
                  help='List all available configs (default)')

agrp.add_argument('-s', '--set', metavar='config',
                  help='Set a config').complete('file', {'directory': NBFC_CONFIGS_DIR})

agrp.add_argument('-a', '--apply', metavar='config',
                  help='Set a config and enable fan control').complete('file', {'directory': NBFC_CONFIGS_DIR})

agrp.add_argument('-r', '--recommend', action='store_true',
                  help='List configs which may work for your device')

# =============================================================================
# Command 'set'
# =============================================================================
cmdp = subp.add_parser('set', help='Control fan speed')
agrp = cmdp.add_mutually_exclusive_group(required=True)

agrp.add_argument('-a', '--auto', action='store_const', dest='speed', const=-1,
                  help='Set fan speed to \'auto\'')

agrp.add_argument('-s', '--speed', type=float, dest='speed', metavar='PERCENT',
                  help='Set fan speed to PERCENT')

cmdp.add_argument('-f', '--fan', type=int, metavar='FAN INDEX', default=0,
                  help='Fan index (zero based)').complete('exec', 'nbfc complete-fans')

# =============================================================================
# Command 'wait-for-hwmon'
# =============================================================================
cmdp = subp.add_parser('wait-for-hwmon', help='Wait for /sys/class/hwmon/hwmon* files')

# =============================================================================
# Command 'help'
# =============================================================================
cmdp = subp.add_parser('help', help='Show help')
